File: SwiftProject/SwiftProject/Kex_2_Battle.swift
//
//  Kex_2_Battle.swift
//  SwiftProject
//
//  Created by 周哲輝 on 2019/06/24.
//  Copyright © 2019 周哲輝. All rights reserved.
//

import Foundation

public func AttackEnemy(player: inout Player, enemy: inout Enemy) -> Bool {
    print("\(enemy.Name)Lv.\(enemy.Lv + 1)が現れた！")
    
    while enemy.Hp > 0 {
        print("\(enemy.Name) 残りHP：\(enemy.Hp)")
        print("武器を選択してください（１、攻撃　２、特技　３、魔法）", terminator: "")
        let choice = Int(readLine()!)!
        let getAttackPower = {
            () -> Int in
            return Int.random(in: 0...40) + 60
        }
        let getSkillPower = {
            () -> Int in
            return Int.random(in: 0...100) + 30
        }
        let getMagicPower = {
            () -> Int in
            return Int.random(in: 0...180) + 20
        }
        var attacks = [Attack(), Attack(), Attack()]
        let getPowers = [getAttackPower, getSkillPower, getMagicPower]
        let getHits = [ATTACK_HIT, SKILL_HIT, MAGIC_HIT]
        for i in 0...attacks.count - 1 {
            attacks[i].Damage = getPowers[i]()
            attacks[i].Hit = getHits[i]
        }
        
        var damage = 0
        var isHit = false
        
        switch choice {
        case 1:
            damage = attacks[0].Damage
            isHit = CheckHitOrMiss(hit: attacks[0].Hit - enemy.Flee)
        case 2:
            damage = attacks[1].Damage
            isHit = CheckHitOrMiss(hit: attacks[1].Hit - enemy.Flee)
        case 3:
            damage = attacks[2].Damage
            isHit = CheckHitOrMiss(hit: attacks[2].Hit - enemy.Flee)
        default:
            break
        }
        
        if isHit {
            damage -= enemy.Defense
            damage = damage <= 0 ? 0 : damage
            print("\(damage)のダメージ！")
            enemy.Hp -= damage
        } else {
            print("攻撃を外した！")
        }
        
        print("\(enemy.Name)の攻撃！")
        isHit = CheckHitOrMiss(hit: enemy.Hit)
        if isHit {
            var injury = enemy.Attack - player.Defense
            injury = injury <= 0 ? 0 : injury
            print("\(injury)のダメージ！")
            player.Hp -= injury
            if player.Hp <= 0 {
                player.Hp = 0
                print("あなたは\(enemy.Name)に負けました！ゲームオーバー！")
                return false
            }
        } else {
            print("攻撃を外した！")
        }
        
        print("プレイヤー残りHP：\(player.Hp)")
        enemy.Hp = enemy.Hp <= 0 ? 0 : enemy.Hp
        if enemy.Hp == 0 {
            print("\(enemy.Name)Lv.\(enemy.Lv + 1)を倒した！")
        }
    }
    return true
}

File: rust/src/questions/kex_2_battle.rs
use crate::player::Player;
use crate::enemy::Enemy;
use crate::read_user_input::read_user_input;
use std::io::Write;
use rand::Rng;
use crate::attack::Attack;
use crate::kex_1sub::{ATTACK_HIT, SKILL_HIT, MAGIC_HIT, check_hit_or_miss};

pub fn attack_enemy(player: &mut Player, enemy: &mut Enemy) -> bool {
    println!("{}Lv.{}が現れた！", enemy.name, enemy.lv + 1);

    while enemy.hp > 0 {
        println!("{} 残りHP：{}", enemy.name, enemy.hp);

        let mut reader = String::new();
        print!("武器を選択してください（１．攻撃　２．特技　３．魔法）＞");
        std::io::stdout().flush().unwrap();
        let choice = match read_user_input(&mut reader) {
            Ok(x) => x,
            Err(ex) => {
                eprintln!("エラーが発生しました：{}", ex.to_string());
                return false
            }
        };

        let get_attack_power = || { rand::thread_rng().gen_range(0, 41) + 60 };
        let get_skill_power = || { rand::thread_rng().gen_range(0, 101) + 30 };
        let get_magic_power = || { rand::thread_rng().gen_range(0, 181) + 20 };
        let get_powers: [Box<Fn() -> i32>; 3] = [Box::new(get_attack_power)
            , Box::new(get_skill_power), Box::new(get_magic_power)];
        let mut attacks = [Attack::new(), Attack::new(), Attack::new()];
        let get_hits = [ATTACK_HIT, SKILL_HIT, MAGIC_HIT];

        for x in 0..3 {
            attacks[x].damage = get_powers[x]();
            attacks[x].hit = get_hits[x];
        }

        let mut damage = 0;
        let mut is_hit = false;

        match choice {
            1 => {
                damage = attacks[0].damage;
                is_hit = check_hit_or_miss(attacks[0].hit - enemy.flee);
            },
            2 => {
                damage = attacks[1].damage;
                is_hit = check_hit_or_miss(attacks[1].hit - enemy.flee);
            },
            3 => {
                damage = attacks[2].damage;
                is_hit = check_hit_or_miss(attacks[2].hit - enemy.flee);
            },
            _ => ()
        };

        if is_hit {
            damage -= enemy.defense;
            damage = if damage <= 0 {
                0
            } else {
                damage
            };

            println!("{}のダメージ！", damage);
            enemy.hp -= damage;
        } else {
            println!("攻撃を外した！");
        }

        println!("{}の攻撃！", enemy.name);
        is_hit = check_hit_or_miss(enemy.hit);
        if is_hit {
            let mut injury = enemy.attack - player.defense;
            injury = if injury <= 0 {
                0
            } else {
                injury
            };
            println!("{}のダメージ！", injury);
            player.hp -= injury;
            if player.hp <= 0 {
                player.hp = 0;
                println!("あなたは{}に負けました！ゲームオーバー！", enemy.name);
                return false
            }
        } else {
            println!("攻撃を外した！");
        }

        println!("プレイヤー残りHP：{}", player.hp);
        enemy.hp = if enemy.hp <= 0 {
            0
        } else {
            enemy.hp
        };

        if enemy.hp == 0 {
            println!("{}Lv.{}を倒した！", enemy.name, enemy.lv + 1);
        }
    }
    true
}